packages/cell-toolbar/src/celltoolbartracker.ts

```diff
@@ -0,0 +1,372 @@
+ */
+export class CellToolbarTracker implements IDisposable {
+  constructor(
+    panel: NotebookPanel,
+    commands: CommandRegistry,
+    settings: ISettingRegistry.ISettings | null
+  ) {
+    this._commands = commands;
+    this._panel = panel;
+    this._settings = settings;
+    this._previousActiveCell = this._panel.content.activeCell;
+
+    if (this._settings) {
+      this._onSettingsChanged();
+      this._settings.changed.connect(this._onSettingsChanged, this);
+    }
+
+    const notebookModel = this._panel.context.model;
+    const cells = notebookModel.cells;
+    cells.changed.connect(this.updateConnectedCells, this);
+
+    // Only add the toolbar to the notebook's active cell (if any) once it has fully rendered and been revealed.
+    panel.revealed.then(() => this._onActiveCellChanged(panel.content));
+
+    // Handle subsequent changes of active cell.
+    panel.content.activeCellChanged.connect(this._onActiveCellChanged, this);
+  }
+
+  _onActiveCellChanged(notebook: Notebook): void {
+    const activeCell = notebook.activeCell;
+    if (!activeCell) {
+      return;
+    }
+
+    if (
+      this._previousActiveCell !== null &&
+      this._previousActiveCell !== undefined
+    ) {
+      this._removeToolbar(this._previousActiveCell.model);
+    }
+    this._addToolbar(activeCell.model);
+    this._previousActiveCell = activeCell;
+
+    this._updateCellForToolbarOverlap(activeCell);
+  }
+
+  get isDisposed(): boolean {
+    return this._isDisposed;
+  }
+
+  dispose(): void {
+    if (this.isDisposed) {
+      return;
+    }
+    this._isDisposed = true;
+
+    if (this._settings) {
+      this._settings.changed.disconnect(this._onSettingsChanged, this);
+    }
+
+    const cells = this._panel?.context.model.cells;
+    if (cells) {
+      cells.changed.disconnect(this.updateConnectedCells, this);
+      each(cells.iter(), model => this._removeToolbar(model));
+    }
+
+    this._panel = null;
+  }
+
+  /**
+   * Callback to react to cells list changes
+   *
+   * @param cells List of notebook cells
+   * @param changed Modification of the list
+   */
+  updateConnectedCells(
+    cells: IObservableUndoableList<ICellModel>,
+    changed: IObservableList.IChangedArgs<ICellModel>
+  ): void {
+    const activeCell: Cell<ICellModel> | null | undefined = this._panel?.content
+      .activeCell;
+    if (activeCell === null || activeCell === undefined) {
+      return;
+    }
+
+    if (changed.oldValues.find(m => m === activeCell.model)) {
+      this._removeToolbar(activeCell.model);
+      this._addToolbar(activeCell.model);
+    }
+  }
+
+  private _addToolbar(model: ICellModel): void {
+    const cell = this._getCell(model);
+
+    if (cell) {
+      const { leftMenu } = (this._settings?.composite as any) ?? {};
+
+      const leftMenu_ = leftMenu === null ? [] : leftMenu ?? DEFAULT_LEFT_MENU;
+
+      const toolbar = new CellToolbarWidget(this._commands, leftMenu_);
```

Comment:
Let's reuse the `Toolbar` widget and use the following logic to populate it: https://github.com/jupyterlab/jupyterlab/blob/master/packages/docregistry/src/default.ts#L432

```suggestion
      const toolbar = new Toolbar();
```